### PR TITLE
Webauthn fixes

### LINKED ISSUE #2178: Support FIDO U2F security key in desktop client
As a user i want to be able to use (login and register) my FIDO U2F security key as a second factor in the desktop client and in the mobile apps.  A security key that is registerd on mobile/desktop application should also be usable within the web version and vise versa.

**AC**
- [ ] Allow setting up a security key in Settings > Login 
- [ ] When security key is registered in desktop client it is always registered for https://mail.tutanota.com
- [ ] Authenticate with security key as second factor in desktop client/mobile app
- [ ] Authenticate with the same key in web client on https://mail.tutanota.com
- [ ] Autenticate with key that has been registered on whitelabel domain.

**Task**
* [x] Implement an ApplicationWindow variant that can be opened from the main thread to a website/url and return some result
* [x] Change the 2F settings to send a request to open `mail.tutanota.com/webauthn`to add a key
* [x] Change login process to authenticate on `mail.tutanota.com/webauthn` with second factor to use built-in webauthn implementation.
 
**Development notes:**
~~We are currently migrating from U2F Api to WebAuthn Api because chrome is dropping support for U2F API. We also should use WebAuthn standard if possible. See also: #2140~~

Instead of implementing the device communication in the native part of the application using different libraries we prefer to use the WebAuthn API of the browser or web view. WebAuthn only works in a secure context (https) and the devices must be registered for a certain domain. In our case the devices are registered for the domain https://tutanota.com when adding a security key in the web client on https://mail.tutanota.com.

The desktop/mobile apps are not running in that secure context. We are creating a web view/browser window and serving the application files from the local file system.

We cannot just create such a secure context for https://tutanota.com locally so the idea is to open another browser window/web view which opens a site on https://mail.tutanota.com whenever we need to confirm the second factor and and just use this window to confirm the second factor.

A conflict occurs when the security key was set up on a whitelabel domain. This key cannot be used to authenticate on http://mail.tutanota.com because it is a different domain. We probaly can also not allow to authenticate within the desktop/mobile app for such keys. They have to register the key for the app again, which then will be registed for http://mail.tutanota.com.

**More information about integrating WebAuthn**
https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/
https://developers.yubico.com/WebAuthn/WebAuthn_Developer_Guide/Desktop_and_Mobile_Authentication.html

**Electron spec:**
https://github.com/electron/electron/issues/15404

Same as https://github.com/tutao/tutanota/issues/443 for desktop client.

## Test notes
 - [ ] Regression testing for web client: adding webauthn keys, authentication with webauthn and legacy U2F keys (might need old client build)
 - [ ] Desktop app:
   - [ ] Add U2F key in settings
     - [ ] Can use cancel button
     - [ ] Can add a key
   - [ ] Authenticate with the key:
     - [ ] key from Tutanota domain (priority 1)
     - [ ] key from whitelabel domain (priority 2)
     - [ ] legacy key from Tutanota (priority 3)
     - [ ] legacy key from whitelabel domain (priority 4)
     - [ ] Can cancel manually
     - [ ] Errors are handled (invalid key)
     - [ ] Automatically cancelled and prompt is closed when confirming from another client
   - [ ] Interactions:
     - [ ] keys registered in desktop work in web client
     - [ ] keys registered in web client work in desktop
   - [ ] Check WebDialog window behaviour (on win, mac, linux)
     - [ ] blocks interaction with the parent window
     - [ ] always is parented to the window we're logging in from, even when quickly changing window after opening
     - [ ] can't be moved independently
     - [ ] is invisible in taskbars and other window lists
     - [ ] looks like a tutanota modal otherwise
   - [ ] regression test #2766

### LINKED ISSUE #3914: WebDialog should not have ServiceWorker enabled
**Describe the bug**

Screenshot from 2022-02-23 11-19-32

**To Reproduce**
1. On desktop authenticate with Webauthn
2. Release a new version of the client
3. Authenticate with webauthn again
4. See update popup

**Expected behavior**
There's no update popup and we always load the most recent version of the client in WebDialog.

**Additional context**
ServiceWorker caching might give us unexpected client version and also has other side-effects like update popup.

### PR DESCRIPTION
**Pull request by johnbotris**
fix #3914

also fix #2178 webauthn dialog's non-responsiveness due to initialisation timing

draft because we still don't handle cancelled errors when the webauthn dialog is closed

### CODE REVIEW

**Comment by johnbotris** [comment]
The unhandled exception was a side-effect of the race condition, so it no longer happens now either

File: src/global.d.ts [hunk 1]
```diff
@@ -43,6 +43,6 @@ declare global {
 		/** Set by the server for whitelabel domains. */
 		whitelabelCustomizations: WhitelabelCustomizations | undefined,
 		/** Whether we are running inside dialog of desktop app. */
-		webDialog: true | undefined,
+		nativeAppWebauthn: NativeApp | undefined,
```
Comment (charlag) [inline src/global.d.ts:46]: I specifically wanted to disable sw for all webdialogs in the future

Comment (johnbotris) [reply src/global.d.ts:46]: my suggestion would then be to rename it to `nativeAppWebdialog`; If we ever find other reasons to use webdialogs, I would assume we would still be using this same preload script and we only really need to expose a single `nativeApp`

Comment (charlag) [reply src/global.d.ts:46]: I am in favor of clearly separating normal IPC and webDialog one, I am okay with `nativeAppWebdialog`, just don't want to hardcode webauthn anywhere

File: src/desktop/WebDialog.ts [hunk 2]
```diff
@@ -62,13 +62,17 @@ export class WebDialog<FacadeType extends object> {
 }
 
 export class WebDialogController implements IWebDialogController {
+
```
Comment (charlag) [inline src/desktop/WebDialog.ts:65]: Some random spaces across the file?

File: src/desktop/WebDialog.ts [hunk 3]
```diff
@@ -81,12 +85,18 @@ export class WebDialogController implements IWebDialogController {
 
 		bw.once('ready-to-show', () => bw.show())
 		bw.webContents.on("did-fail-load", () => closeDefer.reject(new Error(`Could not load web dialog at ${urlToOpen}`)))
-		await bw.loadURL(urlToOpen.toString())
 
-		const facade = await this.initRemoteWebauthn<FacadeType>(bw.webContents)
+		// Don't wait for the facade to init here, because that only happens after we call `bw.loadUrl`
```
Comment (charlag) [inline src/desktop/WebDialog.ts:89]: Good catch!

Comment (charlag) [reply src/desktop/WebDialog.ts:89]: Alternatively we could set up listener and send init separately but I think this solution is also good

**Review by charlag — APPROVED** [review]
Besides naming looks good to me! (I didn't try it locally yet tho)

File: src/desktop/WebDialog.ts (hunk 2 above)
Comment (johnbotris) [reply src/desktop/WebDialog.ts:65]: how'd that get there....

**Pull request merged**